packages/data-model/src/readable/data-instance.ts

```diff
@@ -0,0 +1,74 @@
+import { InteropFactory, ReadableDataRegistration } from '..';
+import { ReadableResource } from './resource';
+
+interface ChildInfo {
+  shapeTree: {
+    iri: string;
+    label: string;
+  };
+  count: number;
+}
+
+export class ReadableDataInstance extends ReadableResource {
+  dataRegistration: ReadableDataRegistration;
+
+  children: ChildInfo[];
+
+  constructor(public iri: string, public factory: InteropFactory, public descriptionLang?: string) {
+    super(iri, factory);
+  }
+
+  private async bootstrap(): Promise<void> {
+    await this.fetchData();
+    await this.buildDataRegistration();
+    if (this.descriptionLang) {
+      await this.dataRegistration?.shapeTree?.getDescription(this.descriptionLang);
+      this.children = await this.buildChildrenInfo();
+    }
+  }
+
+  private async buildDataRegistration() {
+    const dataRegistrationIri = `${this.iri.split('/').slice(0, -1).join('/')}/`;
+    this.dataRegistration = await this.factory.readable.dataRegistration(dataRegistrationIri);
+  }
+
+  public static async build(
+    iri: string,
+    factory: InteropFactory,
+    descriptionLang?: string
+  ): Promise<ReadableDataInstance> {
+    const instance = new ReadableDataInstance(iri, factory, descriptionLang);
+    await instance.bootstrap();
+    return instance;
+  }
+
+  get label(): string | undefined {
+    const predicate = this.dataRegistration?.shapeTree?.describesInstance;
+    if (predicate) {
+      return this.getObject(predicate).value;
+    }
+    return undefined;
+  }
+
+  get shapeTree(): { iri: string; label: string } {
+    return {
+      iri: this.dataRegistration!.shapeTree!.iri,
+      label: this.dataRegistration!.shapeTree!.descriptions[this.descriptionLang]!.label
+    };
+  }
+
+  async buildChildrenInfo(): Promise<ChildInfo[]> {
+    return Promise.all(
+      this.dataRegistration!.shapeTree!.references.map(async (reference) => {
```

Comment:
This is a lot of `!` (and `?` in other places). We should think about what is our approach when data inevitably comes broken or is simply unavailable at the time. Will the service throw appropriate errors?

Reply:
I'm going to create an issue where we can discuss the general approach.

Reply:
tracked in #52